scylla/tests/integration/shard_aware_batching.rs

```diff
@@ -0,0 +1,193 @@
+
+        let serialized_values = values
+            .serialized()
+            .expect("Failed to serialize values")
+            .into_owned();
+
+        let (node, shard_id_on_node) = session
+            .shard_for_statement(&prepared_statement, &serialized_values)
+            .expect("Error when getting shard for statement")
+            .expect("Query is not shard-aware");
+        let destination_shard = DestinationShard {
+            node_id: node.host_id,
+            shard_id_on_node,
+        };
+
+        // Typically if lines may come from different places, the `shards` `HashMap` would be behind
+        // a mutex, but for this example we keep it simple.
+        // Create the task that constitutes and sends the batches for this shard if it doesn't already exist
+
+        let sender = channels_for_shards
+            .entry(destination_shard)
+            .or_insert_with(|| {
+                let (sender, receiver) = tokio::sync::mpsc::channel(10000);
+                let prepared_statement = prepared_statement.clone();
+                let session = session.clone();
+
+                let mut scylla_batch =
+                    scylla::batch::Batch::new(scylla::batch::BatchType::Unlogged);
+                scylla_batch.enforce_target_node(&node, &session);
+
+                batching_tasks.push(tokio::spawn(async move {
+                    let mut batches = ReceiverStream::new(receiver)
+                        .chunks_timeout(10, Duration::from_millis(100));
+
+                    while let Some(batch) = batches.next().await {
+                        // Obviously if the actual prepared statement depends on each element of the batch
+                        // this requires adjustment
+                        scylla_batch.statements.resize_with(batch.len(), || {
+                            scylla::batch::BatchStatement::PreparedStatement(
+                                prepared_statement.clone(),
+                            )
+                        });
+
+                        session
+                            .batch(&scylla_batch, &batch)
+                            .await
+                            .expect("Query to send batch failed");
+                    }
+                }));
+                sender
+            });
+
+        sender
+            .send(serialized_values)
+            .await
+            .expect("Failed to send serialized values to dedicated channel");
+    }
+
+    // Let's drop the senders, which will ensure that all batches are sent immediately,
+    // then wait for all the tasks to finish, and ensure that there were no errors
+    // In a production setting these dynamically instantiated tasks may be monitored more easily
+    // by using e.g. `tokio_tasks_shutdown`
+    std::mem::drop(channels_for_shards);
+    for task in batching_tasks {
+        task.await.unwrap();
+    }
+
+    // finally check that batching was indeed shard-aware.
+
+    // TODO
```

Comment:
I'm not exactly sure how to implement this, help would be welcome

Reply:
Please take a look at #788. Does this solve you issue?